src/mpid/ch4/netmod/ofi/ofi_am_impl.h

```diff
@@ -297,11 +297,11 @@ static inline int MPIDI_OFI_am_isend_long(int rank,
     if (MPIDI_OFI_ENABLE_MR_SCALABLE) {
         /* Always allocates RMA ID from COMM_WORLD as the actual associated communicator
          * is not available here */
-        uint64_t index =
+        lmt_info->rma_key =
             MPIDI_OFI_index_allocator_alloc(MPIDI_Global.mr_key_allocator, MPL_MEM_RMA);
-        MPIR_Assert(index < MPIDI_Global.max_huge_rmas);
-        lmt_info->rma_key = MPIDI_OFI_rma_key_pack(comm->context_id,
-                                                   MPIDI_OFI_KEY_TYPE_HUGE_RMA, index);
+        MPIR_ERR_CHKANDSTMT(lmt_info->rma_key >=
+                            (((uint64_t) 1) << (MPIDI_Global.max_mr_key_size * 8)), mpi_errno,
+                            MPI_ERR_OTHER, goto fn_fail, "**ofid_mr_reg");
     } else {
```

Comment:
I do not understand the condition check in `MPIR_ERR_CHKANDSTMT`. Can we have a short comment here explaining the logic ?

Reply:
It looks like this commit just fixes the previous commit . Can we squash these two commits ?

Reply:
The provider may support a key less than 64 bits. Since the id allocator only allocates 64-bit keys (sequentially started from 0), we need to check if the allocated key is within the supported key length. For example, if the provider supports 32-bit key, then any key larger than 0xFFFF is invalid.

Reply:
I will squash them.

Reply:
It is poor practice to simply abort with a valid provider, unless it's documented as a required OFI feature by MPICH.  @raffenet was putting together a list.  It'll need to go in that list if it's really a required feature.

Reply:
1. The allocator *can* allocate 64-bit keys, but we know at init time what the maximum key size is in `MPIDI_Global.max_mr_key_size`. Are we allowing the allocator to produce keys that can't be used?

2. libfabric will return errors for bad keys.
```-FI_ENOKEY
The requested_key is already in use.
-FI_EKEYREJECTED
The requested_key is not available. They key may be out of the range supported by the provider, or the provider may not support user-requested memory registration keys.
```
It seems to me this check can just be removed.

Reply:
I have removed this check and force pushed.